keras_contrib/layers/normalization.py

```diff
@@ -1,3 +1,235 @@
+        self.running_std = self.add_weight(shape, initializer='one',
+                                           name='{}_running_std'.format(self.name),
+                                           trainable=False)
+
+        self.r_max = K.variable(np.ones((1,)), name='{}_r_max'.format(self.name))
+
+        self.d_max = K.variable(np.zeros((1,)), name='{}_d_max'.format(self.name))
+
+        self.t = K.variable(np.zeros((1,)), name='{}_t'.format(self.name))
+
+        if self.initial_weights is not None:
+            self.set_weights(self.initial_weights)
+            del self.initial_weights
+        self.built = True
+
+    def call(self, x, mask=None):
+        if self.mode == 0 or self.mode == 2:
+            assert self.built, 'Layer must be built before being called'
+            input_shape = K.int_shape(x)
+
+            reduction_axes = list(range(len(input_shape)))
+            del reduction_axes[self.axis]
+            broadcast_shape = [1] * len(input_shape)
+            broadcast_shape[self.axis] = input_shape[self.axis]
+
+            if K.backend() == 'tf':
+                mean_batch, var = K.KTF.nn.moments(x, reduction_axes, shift=None, name=None, keep_dims=False)
+                std_batch = (K.sqrt(var + self.epsilon))
+            else:
+                mean_batch = K.mean(x, axis=reduction_axes)
+                std_batch = K.sqrt(K.var(x, axis=reduction_axes) + self.epsilon)
+
+            r_max_val = K.get_value(self.r_max)
+            r = std_batch / self.running_std
+            r = K.stop_gradient(K.clip(r, 1 / r_max_val, r_max_val))
+
+            d_max_val = K.get_value(self.d_max)
+            d = (mean_batch - self.running_mean) / self.running_mean
+            d = K.stop_gradient(K.clip(d, -d_max_val, d_max_val))
+
+            if sorted(reduction_axes) == range(K.ndim(x))[:-1]:
+                x_normed_batch = (x - mean_batch) / std_batch
+                x_normed = (x_normed_batch * r + d) * self.gamma + self.beta
+            else:
+                # need broadcasting
+                broadcast_mean = K.reshape(mean_batch, broadcast_shape)
+                broadcast_std = K.reshape(std_batch, broadcast_shape)
+                broadcast_r = K.reshape(r, broadcast_shape)
+                broadcast_d = K.reshape(d, broadcast_shape)
+                broadcast_beta = K.reshape(self.beta, broadcast_shape)
+                broadcast_gamma = K.reshape(self.gamma, broadcast_shape)
+
+                x_normed_batch = (x - broadcast_mean) / broadcast_std
+                x_normed = (x_normed_batch * broadcast_r + broadcast_d) * broadcast_gamma + broadcast_beta
+
+            # explicit update to moving mean and standard deviation
+            self.add_update([K.moving_average_update(self.running_mean, mean_batch, self.momentum),
+                             K.moving_average_update(self.running_std, std_batch ** 2, self.momentum)], x)
+
+            # update r_max and d_max
+            t_val = K.get_value(self.t)
```

Comment:
Hmm I guess you are right about it getting to its max value very quickly. The problem is I cant obtain the iteration count from the layer. Or even the epoch count. The paper suggests 40k iterations and 25k iterations for r_max and d_max, but since iteration count is also dependent on batch size, I can't really do much.

While I do wan't to follow the paper to a t, it isnt quite possible yet with Keras working as it is. Also, while it is quite weird that in just 5 iterations it is hitting its max value, this approach seems to have similar results to what the paper describes. Without this increment, it hits nan's due to exploding gradients. With such a sharp increase in its max, it still works. 

How about setting the user adjustable parameter to a default value of 0.1 and multiply it with t_val. That would require roughly 50 iterations instead of 5 to reach its final value. I can't test the effects just yet, since my laptop crashed and my nearly finished training of the BRn WRN-16-4 failed so I have to wait for it to complete again (4-5 hours).